src/local_scheduler/local_scheduler.cc

```diff
@@ -712,10 +743,22 @@ void reconstruct_put_task_update_callback(Task *task,
         /* (2) The current local scheduler for the task is dead. The task is
          * lost, but the task table hasn't received the update yet. Retry the
          * test-and-set. */
+#if !RAY_USE_NEW_GCS
         task_table_test_and_update(state->db, Task_task_id(task),
                                    current_local_scheduler_id, Task_state(task),
                                    TASK_STATUS_RECONSTRUCTING, NULL,
                                    reconstruct_put_task_update_callback, state);
+#else
+        RAY_CHECK_OK(gcs::TaskTableTestAndUpdate(
+            &state->gcs_client, Task_task_id(task), current_local_scheduler_id,
+            Task_state(task), SchedulingState_RECONSTRUCTING,
+            [task, user_context](gcs::AsyncGcsClient *client,
+                                 const ray::TaskID &id, const TaskTableDataT &t,
+                                 bool updated) {
+              reconstruct_put_task_update_callback(task, user_context, updated);
```

Comment:
I feel like this giant statement (appears elsewhere in the PR) warrants a comment.

For instance, one thing that's not immediately obvious to me: is this callback fired after the async TaskTableTestAndUpdate *returns* (my understanding), or after it *finishes* (i.e., accepted by the GCS)?  If it's former, is there a correctness issue?  If this is just for illustrating the new GCS API, a comment explaining this will be nice!

Reply:
The callback is fired after the TaskTableTestAndUpdate *finishes*, so this code does the same thing as the old code and is correct. I will document this behavior in tables.h.

Reply:
@concretevitamin, when you say "returns" are you referring to the "ack" sent by the GCS? In the case with only one replica, that the "ack" is the same thing as the call being "finished". There is no distinction between the two cases.